---
title: "Phylogenetic Tree tutorial"
author: "BiteSize Bioinformatics Session based on a tutorial from  MolecularEcology.com"
date: "6 December 2017"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
setwd("/Users/telatin/Downloads/mammals/")
```

## Molecular ecology tutorial
This document follows the tutorial from http://www.molecularecologist.com/2016/02/quick-and-dirty-tree-building-in-r/.

## Loading libraries
If one or more libraries are not installed, yet, it's possible to install them with the `install.packages("packagename")` command.

We are going to use the **ape**, **phangorn** and **seqinr** packages:

```{r libraries}
# Load libraries
library(ape)
library(phangorn)
library(seqinr)
```

## Alignment File Formats and Conversion 
Using the `read.dna()` function in the package ape, you’ll import your sequence data, choosing between “*interleaved*,” “*sequential*”, “*clustal*” and “*fasta*” formats. 

From there, you’ll want to convert your alignment to a “*phyDat*” object for use in phangorn, the major package for treebuilding in R. 
Once it’s imported, it’s easy to subset your data and only include, say, the first ten taxa:

```{r mammals01}
# Import data [mammals.dna=https://pastebin.com/8Mt3QUTV]
mammals <- read.dna("~/Downloads/mammals/TMEM55A.dna", format="interleaved")
mammals_phyDat <- phyDat(mammals, type = "DNA", levels = NULL)
mammals_phyDat
```

```{r mammalsSubset}
#Subset (first ten)
mammals10 <- subset(mammals_phyDat, 1:10)
```

This function transforsm several DNA formats into the `phyDat` format (package: phangorn).
```{r phydat}
mammals10_phyDat <- phyDat(mammals10, type = "DNA", levels = NULL)
```


## Model Testing and Distance Matrices

Reducing sequence alignments into a matrix of pairwise distances allows for the rapid estimation of phylogenetic trees (at the cost of the additional information those sequences contain). 

To produce a **distance matrix**, however, you’ll need to decide on the model of nucleotide (or protein) evolution that best fits your data, performing a likelihood ratio test.


```{r mammals02}
mt <- modelTest(mammals10)
print(mt)
dna_dist <- dist.ml(mammals10, model="JC69")
```

## Neighbor Joining, UPGMA, and Maximum Parsimony

Once you have a distance matrix, phangorn provides simple, quick functions for *estimating trees* from distance matrices using neighbor-joining and UPGMA algorithms, which can be visualized using the `plot()` function:

```{r mammals03}
mammals_UPGMA <- upgma(dna_dist)
mammals_NJ  <- NJ(dna_dist)
plot(mammals_UPGMA, main="UPGMA")
plot(mammals_NJ, main = "Neighbor Joining")
```

But which of these trees is a better fit for your data? 

Using the `parsimony()` function, you can compare their respective parsimony scores. The function `optim.parsimony()` lets you go a step further, searching treespace through nearest-neighbor interchange (NNI) and subtree pruning and regrafting (SPR). Alternatively, `pratchet()` allows you to perform the search with the parsimony ratchet algorithm.


```{r parsimony}
parsimony(mammals_UPGMA, mammals10_phyDat)
parsimony(mammals_NJ, mammals10_phyDat)
mammals_optim <- optim.parsimony(mammals_NJ, mammals10_phyDat)
mammals_pratchet <- pratchet(mammals10)
plot(mammals_optim)
plot(mammals_pratchet)
```

## Maximum Likelihood and Bootstrapping

Though more computationally intensive than distance-matrix methods, building trees with maximum likelihood methods allows you to include the full data from your sequence alignment in a statistical framework that estimates model parameters. 

You can begin by computing the likelihood of a given tree with the function `pml()`. Then, the function `optim.pml()` will optimize tree topology and branch length for your selected model of nucleotide evolution.


```{r maximumlikelihood}
fit <- pml(mammals_NJ, mammals10)
print(fit)
fitJC <- optim.pml(fit, model = "JC", rearrangement = "stochastic")
logLik(fitJC)
bs <- bootstrap.pml(fitJC, bs=100, optNni=TRUE, multicore=TRUE, control = pml.control(trace=0))
plotBS(midpoint(fitJC$tree), bs, p = 50, type="p")
```

## Exporting Trees

Finally, unless you’re analyzing relationships among a very limited number of taxa, the system viewer in R is probably going to be insufficient for viewing and saving your tree. The function `write.tree()` allows you to export the output of your analyses in Newick format, useful for further tinkering in programs such as FigTree.

```{r exporting}
write.tree(bs, file="/Users/telatin/Downloads/mammals/bootstrap_example.tre")
```

